pkg/publisher/publisher.go

```diff
@@ -0,0 +1,446 @@
+				}
+			}(chunk),
+		})
+	}
+
+	ctx1, cancel1 := context.WithCancel(ctx)
+	defer cancel1()
+
+	p1 := wpool.New(concurrency, rate.Every(rateLimit))
+	go p1.GenerateFrom(jobs1)
+	go p1.Run(ctx1)
+	for r := range p1.Results() {
+		if r.Err != nil {
+			return fmt.Errorf("executing job %d, %s: %v", r.ID, r.Desc, r.Err)
+		}
+		log.Default().Printf("processed job %d: %s, %v", r.ID, r.Desc, r.Value)
+	}
+	_ = os.RemoveAll(tmpDir)
+	return nil
+}
+
+// CidToCarStorage writes a car file from a cid and uploads it to car storage service.
+func (p *Publisher) CidToCarStorage(ctx context.Context, c cid.Cid) (cid.Cid, error) {
+	carReader, carWriter := io.Pipe()
+
+	jobs := []wpool.Job{
+		{
+			ID:   1,
+			Desc: "write car",
+			ExecFn: func(ctx context.Context) (interface{}, error) {
+				if err := car.WriteCar(ctx, p.ipfsClient.Dag(), []cid.Cid{c}, carWriter); err != nil {
+					return nil, fmt.Errorf("writing car: %v", err)
+				}
+				_ = carWriter.Close()
+				return nil, nil
+			},
+		},
+		{
+			ID:   2,
+			Desc: "upload",
+			ExecFn: func(ctx context.Context) (interface{}, error) {
+				c, err := p.carStorage.PutCar(ctx, carReader)
+				if err != nil {
+					return nil, fmt.Errorf("uploading car: %v", err)
+				}
+				return c, nil
+			},
+		},
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	pool := wpool.New(2, rate.Inf)
+	go pool.GenerateFrom(jobs)
+	go pool.Run(ctx)
+	var res cid.Cid
+	for r := range pool.Results() {
+		if r.Err != nil {
+			return cid.Cid{}, fmt.Errorf("executing job %d, %s: %v", r.ID, r.Desc, r.Err)
+		}
+		if r.ID == 2 {
+			res = r.Value.(cid.Cid)
+		}
+	}
+	log.Default().Printf("%s uploaded to car storage service", res.String())
+	return res, nil
+}
+
+// RigsIndexToCarStorage creates a dagcbor index from the provided Rigs and adds it to car storage service.
+func (p *Publisher) RigsIndexToCarStorage(ctx context.Context, rigs []local.Rig) (cid.Cid, error) {
+	n, err := qp.BuildMap(basicnode.Prototype.Any, int64(len(rigs)), func(ma datamodel.MapAssembler) {
+		for _, rig := range rigs {
+			c, err := cid.Decode(*rig.RendersCid)
+			if err != nil {
+				log.Default().Printf("error decoding rig cid string: %v", err)
+				continue
+			}
+			qp.MapEntry(ma, fmt.Sprintf("%d", rig.ID), qp.Link(cidlink.Link{Cid: c}))
+		}
+	})
+	if err != nil {
+		return cid.Cid{}, fmt.Errorf("building map: %v", err)
+	}
+
+	reader, writer := io.Pipe()
+	jobs := []wpool.Job{
+		{
+			ID:   1,
+			Desc: "encode ipld",
+			ExecFn: func(ctx context.Context) (interface{}, error) {
+				if err := dagcbor.Encode(n, writer); err != nil {
+					return nil, fmt.Errorf("encoding ipld: %v", err)
+				}
+				_ = writer.Close()
+				return nil, nil
+			},
+		},
+		{
+			ID:   2,
+			Desc: "upload ipld",
+			ExecFn: func(ctx context.Context) (interface{}, error) {
+				c, err := p.carStorage.PutCar(ctx, reader)
+				if err != nil {
+					return nil, fmt.Errorf("uploading ipld: %v", err)
+				}
+				return c, nil
+			},
+		},
+	}
```

Comment:
This code spins off two concurrent jobs. One for encoding map as dagcbor and the other to upload the resulting data to web3.storage/nft.storage. `p.carStorage` is an interface type that abstracts both web3.storage and nft.storage implementations so I can easily switch between the two. Job 2 gives me an `Unexpected EOF` error when using web3.storage and an `Unexpected end of data` error when using nft.storage. Obviously I'm doing somthing wrong, but can't see what.